hathor/ipykernel.py

```diff
@@ -0,0 +1,64 @@
+# Copyright 2024 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from logging import getLogger
+from typing import TYPE_CHECKING, Any, Optional
+
+from ipykernel.kernelapp import IPKernelApp as OriginalIPKernelApp
+
+if TYPE_CHECKING:
+    from hathor.manager import HathorManager
+
+
+class IPKernelApp(OriginalIPKernelApp):
+    def __init__(self, runtime_dir: Optional[str] = None):
+        super().__init__()
+        if runtime_dir is not None:
+            self.connection_dir = runtime_dir
+        self.connection_file = 'kernel.json'
+        self.logging_config: dict[str, Any] = {}  # empty out logging config
+        self.default_extensions: list[str] = []  # disable default extensions
+        self.no_stderr = True  # disable forwarding of stderr (because we use it for logging)
+        self.log = getLogger('hathor.ipykernel')
+
+    # ignore registering of signals
+    def init_signal(self) -> None:
+        pass
+
+    # custom logging of connection file
+    def log_connection_info(self) -> None:
+        self.log.info(f'ipykernel connection enabled at {self.abs_connection_file}')
+
+    # disable original logging setup
+    def get_default_logging_config(self) -> dict[str, Any]:
+        return {"version": 1, "disable_existing_loggers": False}
+
+    def configure_tornado_logger(self) -> None:
+        pass
+
+    def start(self):
+        self.kernel.start()
+
+
+def embed_kernel(manager: 'HathorManager', *,
+                 runtime_dir: Optional[str] = None, extra_ns: dict[str, Any] = {}) -> None:
+    # get the app if it exists, or set it up if it doesn't
+    if IPKernelApp.initialized():
```

Comment:
Not in practice, but it could be in theory. I've tried to preserve the original behavior of allowing that when making this function (which mimics `embed_kernel` in `IPython`, that just forwards to `ipykernel`). In the future, if we decide to allow starting/stopping the kernel through sysctl, this is the function that would be used to start it, and then it would be called more than once.

Reply:
Good points. Thanks!